cmd/console-proxy/src/posix.rs

```diff
@@ -0,0 +1,401 @@
+    core: &'a mut dyn Core,
+    context: HiffyContext<'a>,
+    poll_interval: Duration,
+}
+
+impl<'a> UartConsoleHandler<'a> {
+    pub fn new(
+        hubris: &'a HubrisArchive,
+        core: &'a mut dyn Core,
+        hiffy_timeout: u32,
+        poll_interval: u32,
+    ) -> Result<Self> {
+        let context = HiffyContext::new(hubris, core, hiffy_timeout)?;
+        Ok(Self {
+            hubris,
+            core,
+            context,
+            poll_interval: Duration::from_millis(u64::from(poll_interval)),
+        })
+    }
+
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let op = IdolOperation::new(
+            self.hubris,
+            "ControlPlaneAgent",
+            "uart_read",
+            None,
+        )
+        .context("Could not find `ControlPlaneAgent.uart_read` operation")?;
+
+        let value = humility_cmd_hiffy::hiffy_call(
+            self.hubris,
+            self.core,
+            &mut self.context,
+            &op,
+            &[],
+            Some(HiffyLease::Read(buf)),
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let op = IdolOperation::new(
+            self.hubris,
+            "ControlPlaneAgent",
+            "uart_write",
+            None,
+        )
+        .context("Could not find `ControlPlaneAgent.uart_write` operation")?;
+
+        let buf = &buf[..usize::min(buf.len(), HIFFY_BUF_SIZE)];
+
+        let value = humility_cmd_hiffy::hiffy_call(
+            self.hubris,
+            self.core,
+            &mut self.context,
+            &op,
+            &[],
+            Some(HiffyLease::Write(buf)),
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn attach(
+        &mut self,
+        raw: bool,
+        imap: RemapRules,
+        omap: RemapRules,
+        log: Option<File>,
+    ) -> Result<()> {
+        // Put terminal in raw mode, if requested, with a guard to restore it.
+        let _guard = if raw {
+            Some(UnrawTermiosGuard::make_stdout_raw()?)
+        } else {
+            None
+        };
+
+        let (stdin_tx, stdin_rx) = crossbeam_channel::unbounded();
+        let (stdout_tx, stdout_rx) = crossbeam_channel::unbounded();
+        thread::spawn(move || stdin_reader(raw, stdin_tx, omap));
+        thread::spawn(move || stdout_writer(stdout_rx, imap, log));
+
+        let mut rx_buf = vec![0; HIFFY_BUF_SIZE];
+        let mut tx_buf = Vec::new();
+        loop {
+            if !tx_buf.is_empty() {
```

Comment:
Would there be any advantage to looping here until `tx_buf` is empty, rather than only doing one write?

Reply:
Yes, although I'm slightly worried about deadlocking (e.g., if we're trying to send until `tx_buf` is empty, but the console uart rx fifo is full and blocked waiting for us to receive). Maybe that's not possible with a well behaving host; not sure.

Alternatively: there's a check below to shorten the sleep time if we just had a full-length read; I could make that "if we had a full-length read or we still have data to transmit". I think that would be nearly equivalent to adding the loop here, except that we pause momentarily to check for incoming data after each write.

Reply:
Agreed, the latter seems like a good idea!